File: exercises/leetcode/algorithms/2558-take-gifts-from-the-richest-pile/solution.ts
function pickGifts(gifts: number[], k: number): number {
   const maxHeap = gifts.map((gift) => -gift).sort((a, b) => a - b)

   for (let i = 0; i < k; i++) {
      // Remove the first element (the highest negative value) from the array,
      // making it positive again to get the original gift count.
      let current = -maxHeap.shift()

      // Push the remaining number of gifts (after taking some) back into the array
      // as a negative value.
      maxHeap.push(-Math.floor(Math.sqrt(current)))

      // Re-sort the array to maintain the heap property.
      maxHeap.sort((a, b) => a - b)
   }

   // Calculate the sum of the remaining gifts. The use of `-` in front of the
   // reduce function inverts the negative values while summing them up.
   return -maxHeap.reduce((a, b) => a + b, 0)
}

File: exercises/leetcode/algorithms/2558-take-gifts-from-the-richest-pile/solution.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        max_heap = [-gift for gift in gifts]

        # Convert the list into a heap in-place. This operation ensures the list maintains heap properties.
        heapq.heapify(max_heap)

        # For each of the 'k' operations, select and process the pile with the most gifts.
        for _ in range(k):
            # Using heappop(), we retrieve and remove the smallest number. Since we've negated the numbers,
            # this effectively gives us the largest original number (i.e., the pile with the most gifts).
            current = -heapq.heappop(max_heap)

            # Calculate the number of gifts left after taking from the pile and push the negative
            # of that value into the heap.
            heapq.heappush(max_heap, -int(current ** 0.5))

        # Sum up the remaining gifts in all the piles. Since the values in the max_heap are negative,
        # we negate the sum to get the final positive total of remaining gifts.
        return -sum(max_heap)
